### PR TITLE
bug: fix_memory_cost_calculation

### PR DESCRIPTION
**Pull request by TAdev0**
## Pull Request type

fix wrong memory cost calculation

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #700 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot-ssj/701)

### CODE REVIEW

**Comment by TAdev0** [comment]
Sure. Will do that in a few hours

File: crates/evm/src/gas.cairo [hunk 1]
```diff
@@ -118,7 +118,7 @@ fn calculate_message_call_gas(
 ///  x = 512 q0 + r0 => x^2 = 512(512 q0^2 + q0 r0) + r0^2
```
[the remark this answers is not in the document]
Comment (TAdev0) [reply crates/evm/src/gas.cairo:118]: done

**Comment by TAdev0** [comment]
@enitrat  PR updated. 
Indeed, tests first failed because  `words_len` was the result of `bytes_len/32` , leading to truncation error. Tests now pass with `(bytes_len + 31) / 32`, all good!

File: crates/evm/src/tests/test_gas.cairo [hunk 2]
```diff
@@ -38,7 +41,7 @@ fn test_calculate_intrinsic_gas_cost() {
     let expected_cost: u128 = 21136;
     let out_cost: u128 = calculate_intrinsic_gas_cost(@tx);
 
-    assert(out_cost == expected_cost, 'wrong cost');
+    assert!(out_cost == expected_cost, "wrong cost");
```
Comment (enitrat) [inline crates/evm/src/tests/test_gas.cairo:44]: assert_eq

File: crates/evm/src/tests/test_gas.cairo [hunk 3]
```diff
@@ -78,7 +81,7 @@ fn test_calculate_intrinsic_gas_cost_with_access_list() {
     let expected_cost: u128 = 21136 + ACCESS_LIST_ADDRESS + 5 * ACCESS_LIST_STORAGE_KEY;
     let out_cost: u128 = calculate_intrinsic_gas_cost(@tx);
 
-    assert(out_cost == expected_cost, 'wrong cost');
+    assert!(out_cost == expected_cost, "wrong cost");
```
Comment (enitrat) [inline crates/evm/src/tests/test_gas.cairo:84]: assert_eq!

File: crates/evm/src/tests/test_gas.cairo [hunk 4]
```diff
@@ -110,5 +113,13 @@ fn test_calculate_intrinsic_gas_cost_without_destination() {
     let expected_cost: u128 = 53138;
     let out_cost: u128 = calculate_intrinsic_gas_cost(@tx);
 
-    assert(out_cost == expected_cost, 'wrong cost');
+    assert!(out_cost == expected_cost, "wrong cost");
```
Comment (enitrat) [inline crates/evm/src/tests/test_gas.cairo:116]: assert_eq!

File: crates/evm/src/tests/test_gas.cairo [hunk 5]
```diff
@@ -110,5 +113,13 @@ fn test_calculate_intrinsic_gas_cost_without_destination() {
...
+}
+
+#[test]
+fn test_calculate_memory_allocation_cost() {
+    let size_in_bytes: usize = 10018613;
+    let expected_cost: u128 = 192385220;
+    let out_cost: u128 = calculate_memory_gas_cost(size_in_bytes);
+    assert!(out_cost == expected_cost, "wrong cost");
```
Comment (enitrat) [inline crates/evm/src/tests/test_gas.cairo:124]: assert_eq!

**Review by enitrat — COMMENTED** [review]
nice catch!

**Comment by TAdev0** [comment]
@enitrat  ready to merge

**Review by enitrat — APPROVED** [review]

**Pull request merged**